flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -266,9 +266,46 @@ class ExprLowering {
         [](const auto &box) -> T { return box; });
   }
 
+  fir::ExtendedValue
+  genAllocatableOrPointerUnbox(Fortran::semantics::SymbolRef sym) {
+    auto boxAddr = symMap.lookupSymbol(sym).getAddr();
+    if (Fortran::semantics::IsAssumedRankArray(sym))
+      TODO("Assumed rank allocatables or pointers");
+    if (Fortran::semantics::IsPointer(sym))
+      TODO("pointer"); // deal with non contiguity;
+    int rank = sym->Rank();
+    // TODO: only intrinsic types other than CHARACTER
+    if (!boxAddr)
+      TODO("Allocatable type not lowered yet");
+    auto boxType = fir::dyn_cast_ptrEleTy(boxAddr.getType());
+    if (!boxType || !boxType.isa<fir::BoxType>())
+      llvm_unreachable("bad allocatable or pointer type in symbol map");
+
+    auto varAddrType = boxType.cast<fir::BoxType>().getEleTy();
+    auto loc = getLoc();
+
+    auto box = builder.create<fir::LoadOp>(loc, boxAddr);
+    auto addr = builder.create<fir::BoxAddrOp>(loc, varAddrType, box);
+    if (rank == 0)
+      return addr;
+    auto idxTy = builder.getIndexType();
+    llvm::SmallVector<mlir::Value, 4> lbounds;
+    llvm::SmallVector<mlir::Value, 4> extents;
+    for (int dim = 0; dim < rank; ++dim) {
```

Comment:
Is `rank` guaranteed to be an `int`? (Often the front-end rigidly uses `int64_t`, etc.)

Reply:
From what I have seen yes, the front-end is pretty consistent using `int` for the rank: https://github.com/llvm/a4a4c503708f8bbf539147187faf180a80b4eea9/blob/master/flang/include/flang/Semantics/symbol.h#L604
Rank cannot be more than 15 I think in Fortran, so it sounds pretty safe to me.
I can also change that to `for (decltype(rank) dim = 0; dim < rank; ++dim)` to avoid any assumptions. Would you prefer that ?